### ISSUE TITLE
Validate claims with jwks file

### ISSUE THREAD

**Issue by idefixcert** [square/go-jose#366]
I try to Validate claims via a JWKS file:

I created an RSA jwks file with private and public key.
when I then try to run 
`err = parsedToken.Claims(keys)`
I get the following error
`square/go-jose: unsupported key type/format`

In my opinion the error is that it is not returned the jose.JSONWebKey in this following line, it is returned the .Key thats the `rsa.PrivateKey`

https://github.com/square/go-jose/blob/c9ac459e06bc9a5ab7efd5d3b3f421d6de27401f/jwt/jwt.go#L168

I created a Sample program for this:

```
package main

import (
	"encoding/json"
	"fmt"
	"gopkg.in/square/go-jose.v2"
	"gopkg.in/square/go-jose.v2/jwt"
	"os"
	"strings"
	"time"
)

var (
	jwks = `{
    "keys": [
        {
            "p": "zQRQEBSVCo0R95Vsx09PnNGdv_pSZByEI9hVkmIOaNAksQxRPWFUS0hBllVZeeSwmejqSyVBwY-SMW7WRfzm8Z_MJNtZCMK1ckpCuXOa8p60eZtv92RGuULFFE2VQlv9Dw8SluXDiBhwXzssoWc5TL8loy0mxJOsCrHcp4bp6hE",
            "kty": "RSA",
            "q": "yHW_W_qituHSDNjib-GtrjiLSRrCPdkpIJWYl5PB5J9cMhWCk6g2wv8KaA68WlXjrUF1jLnzhi34HsRDFxzAHgv7G6aSsJuON2zfQ9HJVxWloUmVC9-IqEHyd6xcaB5KrpgQ4Bs5ETQUHTyIiNFX3kvqmc8dZMKAoXXLKKZu5G0",
            "d": "eXRt3C9oEiEktIdf8HuFC5WViSrY1ihkipZyp1FHiFtDSlJaj-Dkq7UJI_uoOQuuuBbRvilMYYx9xTVdRjSP423KFK1ZJ0fWI02CREOSEQ_klpzVJfS6VfyEGjqZ5tu8D-6YoJbeuFDDNx_6yH3erP-9Y0vdr681f5c206UhLsHTZHVONk0J427pI7BDxetM_Gw3ajMi8YKgfWMR_57X8DMFFk3ro0uOz4-rNwCm-XvRWMNBTWsfMoYVmkhIv54GrEWBaDE3hSyZ3ZCJO4Gqdvkid5gSyL37gKYMyeOimf9XupaouEFzByPh_TKiYar19cKvHw8gEMMeunxI3PLmgQ",
            "e": "AQAB",
            "use": "sig",
            "kid": "test",
            "qi": "s2nTQoBiEZmNu7eIpMdRhJF-dYPsmqUI5vtsI46H4UsDzmComYG6inpifOk2BF6UOyaqamah-KjDpNzSLls1HnYAUiOLH_fynJlEV1n2PBXiT-7tIRRN5KyVC4YD33x8vHM8tM3dASYhBEWDVDRL21IN29mv2RcXxXPYnFfIrjI",
            "dp": "ei40zLCHdidGaueJUWqpxoqB2dSD35fDBF27-aYocjyBcWYm0E693rOsKryou2UX_lF5t9j3TDSU6pLkbgdfVuZl0h1K_VaUrg0loMINuMPuSH7jBmujXms1ankmF8sckCnqaTV7BxORfJ-WZIZQ9rT_6DRTY1micSk4mDpmExE",
            "alg": "RS256",
            "dq": "BICgi7BVd0YbMa6tZ1XqUL90YW_4bOszwUea5cyrzjbGG1ngVPCySi7wTOYH1P-Y3Yta2I5Uqe7eY9KEr0A2EHa0bSqQoObdiuEnIp7TpuEVl0nPzzR2HVFd4NI-QRhtFP7VYjM6VOiWixGQ_4G0N4n8-vdr8ZdszDOZHJCgbQE",
            "n": "oImqxQhq0s7UkNZly9Qe-W-KWyRj1xKh3skKE8jvH_lzh2BWwK-8fVRBw2jC_EBFFPsQKm-umhgOjFnaFFozchqw5xmrndKHuY1NodVX-SdHkGZ_dkGoA1b9TGCNRY5noiByvfCcNmV8F0V5nZftTbKLUV4b4Pazu97dbpBTNYB0NPMhpVkxd0egxkwIRPybHnKJldepXfMqjTzLtH5hpkoiI9C1QWJBysDkShqBuN_RwheHP1aOCkrL88RiIO0jHfUcCy57eWcL245JIOSPuGXX3B7-6VBo38BnY80nkaNaxybzmyY7GuzEbVKWbp87DYD3CJY7CQ2jpNtuX13NPQ"
        }
    ]
}`
)

func main() {
	var keys jose.JSONWebKeySet
	decoder := json.NewDecoder(strings.NewReader(jwks))
	decoder.Decode(&keys)

	//sign
	key := keys.Key("test")[0]
	signingKey := jose.SigningKey{Algorithm: jose.SignatureAlgorithm(key.Algorithm), Key: key}
	signingOptions := (&jose.SignerOptions{}).WithHeader("kid", key.KeyID).WithType("JWT")
	signer, err := jose.NewSigner(signingKey, signingOptions)
	if err != nil {
		fmt.Printf("%v", err)
		os.Exit(1)
	}
	defaultClaims := &jwt.Claims{
		Expiry:   jwt.NewNumericDate(time.Now().Add(time.Minute)),
		IssuedAt: jwt.NewNumericDate(time.Now()),
	}
	token, err := jwt.Signed(signer).Claims(defaultClaims).CompactSerialize()
	if err != nil {
		fmt.Printf("%v", err)
		os.Exit(1)
	}

	parsedToken, err := jwt.ParseSigned(token)
	if err != nil {
		fmt.Printf("%v", err)
		os.Exit(1)
	}
	err = parsedToken.Claims(keys)
	if err != nil {
		fmt.Printf("%v", err)
		os.Exit(1)
	}
}
```

**Comment by nick9822** [comment]
@idefixcert I agree the errors are ambiguous here and that's how I came across this issue.  I noticed couple of things:
- In JWK set each element inside keys represents a single `JSONWebKey`. In your case, your key contains `d` which means it's a RSA private key. So, after decoding your first and only element is deemed as a RSA Private Key even though it has necessary information i.e. `n`, `e` to create a RSA Public key.
- The code snippet you mentioned select the only one `keys[0].Key` the first element of the slice returned by `Key(kid)`. Here, the `Key` method returns all the `JSONWebKey` matching `kid`.  The method has following comments:
> Key convenience method returns keys by key ID. Specification states that a JWK Set "SHOULD" use distinct key IDs, but allows for some cases where they are not distinct. Hence method returns a slice of JSONWebKeys.
- So, basically here you are trying to sign and verify using the same key i.e. your RSA Private key, which obviously won't work.
- You will have to add another key with different kid for RSA Public key of course without `d`, `p`, `q`. The code only checks for `d` for deciding whether it's a RSA Private key.
